### ISSUE TITLE
(🎁) Support `typing.assert_never` (and the one from `typing_extensions`)

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#12399]
```py
def assert_never(__arg: Never) -> Never: 
    """Custom impl for demonstration purposes only. (mypy doesn't have stubs for the real one)"""
    ...

assert_never()  # no error
```

Mypy should report an error when `assert_never` is used in a position that is reachable.

Related: #12253

**Comment by JelleZijlstra** [comment]
Your example gives two straightforward errors (`Never` is not defined and `assert_never()` takes one argument).

This fixed version behaves as expected:

```python
from typing import NoReturn as Never

def assert_never(__arg: Never) -> Never: 
    """Custom impl for demonstration purposes only. (mypy doesn't have stubs for the real one)"""
    ...

def f(x: bool):
    assert_never(x)  # error
    if x is True:
        pass
    elif x is False:
        pass
    else:
        assert_never(x)  # no error
```

Can you clarify what you think is wrong?

`typing.assert_never` shouldn't need any specific support in mypy because the stubs are sufficient.

**Comment by KotlinIsland** [comment]
# 😳
True, sorry about that.

**Comment by KotlinIsland** [comment]
@JelleZijlstra Why does `Never`/`raise` suppress unreachable errors? #10916
